[1단계 - 자동차 경주 구현] 수달(허수진) 미션 제출합니다.

her0807 opened the pull request:
> # 자기소개
안녕하세요. 리뷰어님! 

**수달**  (🦦)입니다.
우테코를 통해 리뷰어님을 만나게 되어 영광입니다. 

피드백 받기에 앞서, 제가 생각하는 코딩 수준은 학습이 많이 필요한 상태라고 느껴집니다.

객체 지향적이란 무엇일까, 어떻게 하면 유지보수하기 좋은 코드가 될까, 이런 저런 생각들이 많지만 
아직 정리가 안되어 있는 상태입니다. 😥 

그러다보니 코드를 작성할 때 우선순위나 중요도를 판단하기가 힘든 것 같습니다. ㅠㅠ

그래서 이번 **피드백의 목표는 : 코드에 소신을 가지자** 로 설정했습니다. 

많이 알려주시면 빠르게 배우겠습니다.  아자 아자!!! 👍🏻 

---


> # 코드 작성할 때 고민되었던 점

## 1. 테스트
-  **테스트 코드를 작성하실 때, 이것만은 지켜라! 하는게 있으신가요?** 
이전에 코딩할 때 테스트 코드를 작성하지 않아서, 이번에 처음 테스트 코드를 작성해보았습니다.  
그래서 단위 테스트가 어떤 것인지, 테스트가 어떤 것인지 정확하게 감이 안잡히는 상황에서 처음 테스트 주도 개발(tdd)로 작성하려다 보니 어려움이 많았습니다.  ㅠㅠ

-  **private 메서드를 테스트하는 방법이 궁금합니다.**
최대한 작은 단위로 테스트를 하고 싶은데, indent 를 고려하다보니 메서드가 작게 쪼개져 있어 public 으로 쪼갠 메서드 들을 동작하게 구현을 해두었습니다.  그러다보니 대부분의 메서드가 private 로 지정이 되어 있어 테스트가 어렵더라구요. 일단 임시 방편으로 메서드를 test class 에 복붙해서 거기서 test 를 시도했는데, 더 좋은 방법이 있을까요?
 `ex -Cars.java 에 hasNext()`
---

## 2.검증
- 데이터  검증을 `입력 받는 클래스` 와 `객체 내부` 둘 중 어디서 하는게 더 바람직할까? 고민이 있었습니다. 
제가 생각하기에는 입력 받자마자 그 값이 유효한지 판단하고, 거기서 에러처리를 하는게 더 시간적으로 효율적이라고 생각이 되어 
입력 받는 클래스에서 구현을 했습니다. 어떤게 더 바람직 하다고 생각하시나요? 😀 


 ## 3.리뷰 받는 사람이 가져야 할 자세
처음 리뷰를 받아서 리뷰를 해주시는 분께 어떤 부분을 설명하고, 질문 드리면, 답변에 도움이 되실 지 궁금합니다.

**ex) 다음에 리뷰를 받을 때는 `"~ 이런 설명이 있으면 좋겠다" `, 혹은 `"~ 것을 중요하게 생각하고 코드를 작성했으면 좋겠다"`**



---

# 귀한 시간 내주셔서 감사합니다.  🤓 
  
![스크린샷 2022-02-11 오후 4 59 37](https://user-images.githubusercontent.com/26570275/153555964-a3c0d7bd-dde8-4283-8dff-1169b1e154cb.png)

her0807 committed:
feat: 계산기 생성

her0807 committed:
test: String에 관한 테스트 추가

her0807 committed:
test: Set에 관한 테스트 추가

her0807 committed:
test: 테스트 케이스를 추가함

her0807 committed:
test: 중간에 공백이나 null 이 들어올 경우

her0807 committed:
test: 값이 문자일 경우 테스트

her0807 committed:
feat: 값이 Null or 빈값일 경우 예외처리

her0807 committed:
feat: 숫자 하나만 들어올 경우 예외 추가

her0807 committed:
feat: , : 구분자 추가

her0807 committed:
feat: 커스텀 구분자가 있을 경우 기능 추가

her0807 committed:
feat: 음수일 경우 예외 추가

her0807 committed:
docs: 기능 구현 목록 추가

kyukong committed:
test: 자동차 이름이 공백일 경우 테스트 추가

kyukong committed:
test: 입력한 자동차 개수가 2개 미만인 경우 테스트 추가

kyukong committed:
test: 자동차 이름이 중복될 경우 테스트 추가

kyukong committed:
test: 자동차 이름의 길이가 5글자를 초과하는 경우 테스트 추가

kyukong committed:
feat: 자동차 이름이 공백인 경우 추가

kyukong committed:
feat: 자동차 개수가 2개 미만인 경우 추가

kyukong committed:
feat: 자동차 이름이 중복될 경우 추가

kyukong committed:
feat: 자동차의 이름이 5글자를 초과하는 경우 추가

kyukong committed:
refactor(Input): 메소드 분리

kyukong committed:
fix: 자동차 입력받는 메소드 수정

- 자동차 이름의 길이 제한 수정(5글자 포함)
- 에러 메시지 출력 추가

kyukong committed:
fix(InputTest): 자동차 입력 테스트 수정

kyukong committed:
refactor: [ERROR] 메시지 추가

kyukong committed:
docs: README 갱신

kyukong committed:
test: 시도횟수 공백 테스트 추가

kyukong committed:
test: 시도횟수 문자 예외 테스트 추가

kyukong committed:
test: 시도횟수 음수 테스트 추가

kyukong committed:
feat: 시도횟수 문자 예외 확인 추가

kyukong committed:
feat: 시도횟수 1미만의 수 확인 추가

kyukong committed:
fix(InputTest): 시도횟수 음수 테스트 시 입력값 타입 수정

- int 타입에서 String 타입으로 수정

kyukong committed:
docs: README 갱신

kyukong committed:
feat: Car 클래스 생성

kyukong committed:
feat: Cars 클래스 생성

kyukong committed:
feat: Attempt 클래스 생성

kyukong committed:
refactor: 데이터 일급객체, 일급컬렉션으로 포장

kyukong committed:
feat: 레이싱 게임을 진행할 RacingCarGame 클래스 생성

kyukong committed:
feat: 자동차 이동 기능 구현

kyukong committed:
feat: 자동차 이동 구현 메소드 실행

kyukong committed:
docs: README 갱신

kyukong committed:
test: 현재 횟수와 시도횟수가 동일한지 확인하는 테스트 추가

kyukong committed:
test: 자동차 위치 이동 테스트 추가

kyukong committed:
feat: 경기 결과 출력

kyukong committed:
docs: README 갱신

kyukong committed:
feat: 최종 우승자 출력

kyukong committed:
refactor: 자동차 위치 이동 테스트 수정

kyukong committed:
fix(Input): 자동차 이름 길이 예외처리 수정

kyukong committed:
refactor: 들여쓰기 수정함에 따라 관련 코드 수정

kyukong committed:
docs: README 갱신

her0807 committed:
refactor(Car): 상수 추출과 메서드 분리

- Car와 Cars에 역할을 명확하게 하기 위해 메서드 이동
- 주요기능인 drive 에 관한 test 수행

her0807 committed:
refactor(Cars): 상수 추출과 자동차 경주, 우승자 추출 로직 테스트

- hasNext() 에 가중된 책임을 generate() 메서드분리로 해결
- 상수 추출
- 랜덤 숫자 범위 test ,우승자 추출 test

her0807 committed:
refactor(RacingCarGame): 조금더 직관적인 네이밍으로 수정, 상수 추출

her0807 committed:
refactor(Input): 상수 추출, 공백과 null error 분리

- 상수선언
- null 과 공백일 경우 에러메세지 상세하게 출력
- 자동차 이름 입력 값에 대한 Null 처리 test 추가
- 반복 횟수 입력 값에 대한 Null 처리 test 추가

heebong commented on src/main/java/racingcar/view/Input.java:
```diff
@@ -0,0 +1,118 @@
+		}
+	}
+
+	public void allCarsValid(String names) {
```
`allCarsValid()`는 같은 클래스의 `carName ()`에서만 쓰이는데 `public`이네요. `private`으로 바꾸는게 어떨까요?

물론 `private`으로 바뀌면 `allCarsValid()`의 테스트클래스는 만들지 못할거에요 (지금은 없지만요)
해당 로직을 테스트코드로 검증하고 싶다면 어떤 방법이 있을까요? 

hint 새로운 객체를 만들어 역할을 할당하면 `public` 함수로 있고 테스트도 짤 수 있을거에요.

her0807 replied:
해주신 리뷰에 대한 생각은 아래 답글에 남겨두었습니다. ㅎㅎ 🙃

heebong commented on src/main/java/racingcar/view/Input.java:
```diff
@@ -0,0 +1,118 @@
+		return names.length >= CAR_LIMIT;
+	}
+
+	public boolean isDuplicated(String[] names) {
```
`public`으로 되어있네요. 아마 테스트코드를 위해 그럴 것 같아요. 
우선 어떤 함수를 `public`으로 만들어야 할까요? 객체 지향에서 중요한 개념은 `캡슐화`, `은닉`이에요. 내가 원하는 함수만 `공개`할 수 있죠. `public`은 `공개`함수를 지정하는거에요. 때문에 불필요한 내용들은 감추는게 맞아요. 그렇다면 `Input` 객체에서 `isDuplicated()`함수는 공개할 중요한 로직일까요?

아니라고 생각하고 private으로 만들고싶은 욕구가 있을겁니다 😄  하지만 테스트를 위해 어쩔 수 없이 public으로 바꿨다고 생각해요. 그럼 다시한번 생각해봐요. 무엇을 테스트해야할까요?
여기서 봐볼게요. 이름을 입력했을 때 중복이면 안됐으면 좋겠고 길이도 5자 이하였으면 좋겠고... 중요한건 `이름을 입력했을 때`에요. 이름을 입력했을 때 실행되는 밸리데이션 함수가 있나요? 있죠. `allCarsValid()` 이쁘게 하나로 모아서 만들어놨어요.
그러면 우리는 `allCarsValid()` 함수를 테스트하면 됩니다. 이 함수가 중복도 걸러주는지, 길이도 보는지 등의 로직을 수행하는지 이 함수를 통해 확인할 수 있답니다. 그럼 `isDuplicated()`와 같은 함수를 private으로 만들 수 있어요.

무엇을 테스트할지, 무엇을 공개할 지 생각하면 좋을 것 같아요. 결국 테스트도 해당 객체를 사용하는 `사용자`니까요. 그러면 어느정도 private 메서드를 테스트하는 방법의 대안도 됐을거라 생각합니다. 😄

her0807 replied:
희봉의 예상처럼 private으로 만들고싶은 욕구가 상당히 컸습니다.. 😄 
메서드를 분리만 생각하고, 클래스를 분리할 생각은 못했습니다. ㅠㅠ

이번에 주신 피드백을 보고, 클래스 분리의 필요성을 깨달았어요 ㅎㅎㅎ
(제가 너무 늦은 시간에 답변을 달아서 혹시 알람이 갈까봐 걱정이 되네요 ㅠㅠ)

그래서 검증을 담당하는 클래스를 만들고,그 객체를 생성해서 검증하는 형식으로 구조를 변경해서
리팩토링을 진행해 보았습니다.

heebong commented on src/main/java/racingcar/view/Input.java:
```diff
@@ -0,0 +1,118 @@
+	}
+
+	public boolean isDuplicated(String[] names) {
+		Set<String> carNames = new HashSet<>(Arrays.asList(names));
```
`Set`으로 중복 관리 👍

heebong commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+		}
+	}
+
+	public void play() {
```
Cars는 public 메서드가 3개밖에 없지만 생성자를 제외한 2가지 모두 테스트하기 어려운 함수입니다. 왜그럴까요?
`play()` 함수는 car들의 `drive()`를 호출합니다. 하지만 단순히 호출하는게 끝이 아니에요. `앞으로 나아가는지 아닌지 여부를 결정`해서 넣어주거든요. 정확히는 `난수를 생성해서 4 이상이면 전진 아니면 정지` 여부를 `결정`합니다. 그래서 복잡해요.
play()를 더 가볍게 바꿀 순 없을까요? `난수를 생성`하는 역할도 빼고, `전진 아니면 정지 여부를 결정` 하는 역할도 빼고요. 그럼 hasNext, generate 모두 `Cars`의 테스트 코드 안에 없어도 될 것 같아요. 오히려 당당히 public 메서드로 테스트를 검증할 수 있을 것 같습니다. 😃 

hint `car.drive()`에게 역할을 주자.
hint 난수생성은 객체로 제외해보자

her0807 replied:
제가 play() 에게 drive가 할 역할까지 무겁게 지어주었군요 ..! Random 라이브러리도 참고해서 수정해보겠습니다!

heebong commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+		return new Car(name.trim());
+	}
+
+	public List<String> findWinners() {
```
Cars의 테스트하기 어려운 `findWinners()`도 살펴봅시다. 이 함수는 `car들의 최대값`을 구하고 `최대값이 같은 car들을 조회`하는 역할을 합니다. 실제로 이 테스트는 어렵지 않아요. 게임이 끝난 `List<Car>`를 준비하면 어떤 값이 나와야하는지 정할 수 있으니까요. 하지만 여기서는 테스트가 어렵습니다. 바로 `게임이 끝난` List<Car>를 Cars에 설정할 수 없기 때문이에요. 그럼 어떻게 할 수 있을까요?

hint List<Car>를 받는 생성자

Cars의 역할이 무엇인지 생각해서 반영하셨으면 좋겠습니다. 😄

her0807 replied:
제가 설계했던 Cars 는 자동차들을 관리해주는 일급컬렉션 이었는데, 
희봉께서 말씀해주신 내용을 살펴본 후, Cars 의 생성자를 다시 보았더니
Cars 에서 자동차를 생성해주고, Car 객체에 대한 검증 등등 관리가 아닌, 다른 부분들도 눈에 보이기 시작하네요 ㅎㅎㅎ 
처음부터 생성된 List 를 받게된다면 역할을 명확하게 할 수 있겠다 생각이 들어 변경했습니다.

heebong commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,82 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class CarsTest {
```
현재 이 테스트는 Cars의 테스트가 아닙니다. 또다른 Cars의 구현이에요 😢  테스트도 해당 객체를 사용하는 사용자라는걸 잊지마시고 변경하셨으면 좋겠습니다.

물론 Cars가 테스트하기 너무 어려운 객체라서 이런 선택을 하셨을 것 같아요. 하지만 Cars에 남긴 코멘트를 해결하면 Cars도 테스트하기 좋은 객체로 변해 CarTest같은 다른 테스트처럼 검증의 역할을 잘 할거라 생각합니다. 💪

heebong commented on src/test/java/study/StringTest.java:
```diff
@@ -0,0 +1,41 @@
+
+public class StringTest {
+
+	@DisplayName("문자열 자르기 테스트")
```
displayName 사용 👍

heebong commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,35 @@
+class CarTest {
+	Car car = new Car("sudal");
+
+	@DisplayName("랜덤 넘버가 4 이상일 경우 전진")
```
테스트 설명과 테스트 내용이 다르네요. 내용을 통일시키면 좋겠습니다 😃

heebong commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,35 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class CarTest {
+	Car car = new Car("sudal");
```
접근제어자 `private`을 추가해요

heebong commented on src/test/java/study/SetTest.java:
```diff
@@ -0,0 +1,47 @@
+    @DisplayName("set 사이즈 확인 test")
+    @Test
+    void sizeValueTest() {
+        assertThat(numbers.size()).isEqualTo(3);
```
`1. 테스트`와 관련된 질문의 답변을 드리도록 하겠습니다. 😄 
* 테스트를 작성할 때 이것만은 지켜라!
객체는 함수를 가지고있죠. 함수는 보통 인풋과 아웃풋이 있어요. 테스트는 그 인풋과 아웃풋을 확인하는거에요. 이 코드에서 `numbers`가 변하지 않는 한 `numbers.size()`는 3이 나오겠죠? `내부 구현이 어떻게 되는진 모르겠고`, 조건과 조건에 해당하는 결과가 정상적으로 나오는지 확인하는게 테스트입니다. 검증하고자 하는 객체의 public 함수가 실행되야하고, `assert` 함수로 그 내용을 확인해야합니다. 테스트에는 꼭 `실행`-`확인`이 있어야합니다. 
객체지향의 특징 중 하나가 `객체에게 메시지를 보낸다`는거에요. 메시지를 보내는게 함수를 호출한다고 생각하시면 됩니다. 테스트도 메시지를 보내는거에요. 그렇게 생각하시고 짜보시면 더 좋아지실겁니다 👍 

* private 메서드를 테스트하는 방법
이건 위에 Input 코멘트와 Cars 코멘트를 참고하시면 답이 될거라 생각합니다. 실은 간단해요. `구현이 잘못됐거나, 테스트할 함수가 아니거나.`
어떤 상황이 닥쳤을 때 어떻게 해결하는게 좋을지 과제를 진행해가며 수달만의 정답을 찾아가길 바래요 😄

her0807 replied:
불변하는 값을 테스트하는게 아니라, `조건에 따른 결과가 바뀔 때 그때 테스트를 진행한다` !!  😄
테스트가 아직은 익숙하지 않아서 어렵지만, 말씀해주신 내용을 바탕으로 꾸준하게 연습해보겠습니다 :)

heebong commented on src/main/java/racingcar/view/Input.java:
```diff
@@ -0,0 +1,118 @@
+		try {
+			Output.getCarName();
+			String inputValue = scan.nextLine();
+			allCarsValid(inputValue);
```
`2.검증`에 대한 답변 드리겠습니다. 😄 
데이터검증을 입력받는곳에서 할지, 객체 내부에서 할지 고민이 된다고 질문 주셨어요. 지금은 입력받는 곳에서만 검증을 하고 있구요. 
한 번 생각해봐요. `데이터 검증은 누구의 역할인가?`
수달이 생각하기에 Car 객체는 이름이 없어도 되거나, 혹은 5글자 이상이어도 상관 없다고 판단되면 Car 객체는 검증을 하지 않는게 맞아요. 대신 Car 객체를 사용하는 다른 객체들은 Car의 이름이 5글자 이하로 꼭 있을거라고 보장할 수 없겠죠. 하지만 그 이름의 특징이 객체가 가진 고유 특성이라고 판단하면 객체가 검증을 해야겠죠.

이렇게 데이터 검증이 누구의 역할인지 생각해보고 답을 내리면 좋을 것 같아요. 검증의 역할이 어디에 있느냐에 따라 어떤 다른 문제가 발생할 수 있는지도 생각해보면 좋을 것 같아요 😃 

`3. 리뷰 받는 사람이 가져야 할 자세`도 같이 답변드릴게요.
지금처럼 작업 중 궁금한 내용을 정리해서 리뷰 요청때 같이 주시면 좋을 것 같습니다. 어떤 고민이 있는지 잘 알수있어서 더 유익한 답변을 드릴 수 있고, 많이 얻어가실 것 같아요.
코드를 작성할 때도 내가 잘 하고있는게 맞나 고민이 될 수 있어요. 그럴땐 우선 `주차 학습 요구사항`을 잘 지키고 있는지 생각하시면 될 것같아요. 선행학습보다 중요한건 지금 배운것을 활용하는 것이니까요. 여지껏 코드를 작성하던 방식과 다른 방식을 배우고 적용하는거니까 당연히 어렵고 어색할 수 있어요. 그러니 이번에 배운 것들은 꼭 소화하겠다!는 마음으로 과제를 하면 많이 성장하실 것 같습니다 👍  열심히 생각하고 과제하며 코드에 소신을 가지길 바랍니다!

her0807 replied:
## 입력 받는 곳에서 검증을 했던 이유
맨 처음 "Input 에서만 Car 와 Attempt 객체를 생성해야지~"  라는 생각으로 구현을 하여 입력 받는 곳에서 검증을 했었는데,
희봉 피드백을 받고 나니까 `헉.. 객체 생성은 어디에서나 할 수 있구나.. 생성될 때마다 객체 특성을 유지하기 위해서는 객체 내부에 검증 로직이 있으면 믿음직한 객체가 되겠다` 는 생각이 들었어요. 넓은 시야를 선물해주신 희봉 감사합니다.. 광명을 찾았어요 😄😄

heebong commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	private int generate() {
+		return (int)(Math.random() * RANDOM_MULTIPLE) % RANDOM_DIVISION;
```
Math.random()과 new Random.nextInt()는 무엇이 다를까요?

hint https://stackoverflow.com/questions/738629/math-random-versus-random-nextintint

advance⬆️ Random과 ThreadLocalRandom은 무엇이 다를까요?
hint https://www.baeldung.com/java-thread-local-random

her0807 replied:
라이브러리를 깊게 살펴보고 선택할 때 명확한 기준을 가지고 해야겠습니다!!! 😢

heebong reviewed (changes requested):
안녕하세요, 수달! 이번에 리뷰를 맡게된 희봉입니다.
첫 과제 하시느라 너무 고생하셨어요 🎉 
커밋 로그도 잘 남기고 요구사항 정리도 너무 잘하셨습니다. 👍  함수 분리도 잘하셔서 길이와 뎁스를 넘지 않았어요. 추가로 함수 이름을 간결하고 명확하게 작성해서 읽기 편했습니다 💯 

몇가지 코멘트를 남겼습니다. 확인 부탁드려요. 질문주신 내용도 코멘트로 답변 남겼습니다. 
혹시 궁금한 내용이 있다면 바로바로 댓글 혹은 DM 주세요 😄

her0807 committed:
refactor: 검증 책임을 객체로 변경하기 위한 테스트 작성

her0807 committed:
refactor: 검증을 하기 위한 클래스 생성

- 시도 횟수에 관한 검증을 책임질 AttemptValidator
- 자동차 컬렉션에 관한 검증을책임질 CarsValidator
- 자동차 이름에 관한 검증을 책임질 CarValidator
- 공통된 검증 로직을 담당할 Validator

her0807 committed:
refactor: Cars 에 play() 책임 간소화 진행을 위한 test 수정

her0807 committed:
refactor: 전진 여부를 결정하는 로직 수정

- 랜덤 넘버 생성을 위한 StepGenerator
- 자동차를 움직 일지 말지 결정할 drive ()

her0807 committed:
refactor: Cars 생성자를 List 값을 받도록 변경

- List<Car> 값을 받아 자동차들에 관한 관리만 담당하게 변경함.
- 테스트를 위해서 Car에 Name 과 Position 을 받는 생성자 추가
- 그에 따라 Input 에 List<Car> 를 생성하는 메서드 추가

her0807 committed:
refactor: 코드 컨벤션에 맞추어 수정

heebong commented on src/test/java/racingcar/view/InputTest.java:
```diff
@@ -0,0 +1,7 @@
+package racingcar.view;
+
+
+class InputTest {
```
사용하지 않는 테스트네요~!

her0807 replied:
헙..!

heebong commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,49 @@
+		assertThat(winners).containsExactly("bong", "su", "dal");
+	}
+
+	private List<Car> check_ranking_if_correct() {
```
테스트를 위한 fixture 👍 
sudal good 😄

heebong commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,49 @@
+
+	@DisplayName("이동거리 최댓값을 가지는 Car 찾기")
+	@Test
+	void cars_max_position_car() {
```
`cars_max_position_car`와 `cars_winners` 테스트 함수가 같은 조건, 같은 함수를 사용하고 있어서 하나의 테스트 함수로 묶어도 좋을 것 같습니다.

her0807 replied:
저도 만들면서 같은 내용인 것 같다고 생각했는데, 합쳐보겠습니다 :)

heebong commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,46 @@
+	private List<Car> cars = new ArrayList<>();
+
+	public Cars(List<Car> names) {
+		CarsValidator validator = new CarsValidator();
```
advance ⬆️  CarsValidator를 인스턴스 변수로 사용해야 할 이유가 있을까요? `carsValid()` 함수를 `static` 함수로 사용하면 어떤 장점이 있을까요? 😄

her0807 replied:
(마침 이번주에 인스턴스 클래스에 관한 특강을 듣고 피드백을 받으니 리팩토링의 필요성을 느낍니다 ^^;)
`검증 객체는 행위만 있기 때문에` 상태 유지를 위해 인스턴스를 생성하지 않아도 된다!  
static 메서드로 사용하게 되면, 객체 생성 없이 참조가 가능하기 때문에 사용성이 증가한다! = `리팩토링 가-즈아`

heebong commented on src/test/java/racingcar/domain/validation/CarValidatorTest.java:
```diff
@@ -0,0 +1,38 @@
+	void car_name_empty(String name) {
+		assertThatThrownBy(() -> {
+			validator.checkBlank(name, "자동차 이름은 공백일 수 없습니다.");
+		}).isInstanceOf(IllegalArgumentException.class);
```
예외의 메시지까지 확인하면 더 좋을  것 같아요 😄 
hint : `hasMessage()`

her0807 replied:
이 부분은 에외 메세지가 중복으로 사용되는 문제가 있겠다 싶어서 작성하지 않았는데, 
`구체적인 메세지까지 확인하는게 테스트의 질을 높일 수 있겠다` 생각되어 예외 메세지 부분을 
상수 클래스로 추출하여 리팩토링 진행했습니다 😄

heebong commented on src/main/java/racingcar/service/StepGenerator.java:
```diff
@@ -0,0 +1,12 @@
+public class StepGenerator {
+	private static final int RANDOM_RANGE = 10;
+
+	public int generate() {
```
advance⬆️ ` StepGenerator`의 `generate()` 함수가 `static` 함수면 어떤점이 달라질까요? 😃

heebong commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,79 @@
+		}
+	}
+
+	public boolean hasNext(int random) {
```
`hasNext`함수가 정말 `Car`에 public으로 필요한 함수일까요? 테스트를 위한 함수가 아닐까요?
만약 `hasNext()`함수를 `private`으로 변경하면 `drive()` 함수를 테스트해야하는데, 그건 쉽지 않아요. 어떻게하면 테스트하기 쉬운 함수로 변경할 수 있을까요?
hint `stepGenerator.generate()`를 밖으로 빼보기

heebong reviewed (approved):
안녕하세요, 수달! 리뷰어 희봉이에요 🚗 
피드백 반영 잘 해주셨습니다 💯 훨씬 코드가 깔끔해졌어요 👍

1단계는 여기서 머지하도록 하겠습니다. 추가로 적은 코멘트는 2단계때  반영 부탁드립니다. 고생하셨어요 😄
